services/server/src/server/services/storageServices/EtherscanVerifyAPIService.ts

```diff
@@ -0,0 +1,586 @@
+      result[chain] = chainProperties.etherscanApi.apiKeyEnvName;
+    }
+  }
+  return result;
+};
+
+interface EtherscanRpcResponse {
+  status: "0" | "1";
+  message: string;
+  result: string;
+}
+
+export interface EtherscanVerifyAPIServiceOptions {
+  /** Mapping of chainId to the explorer API base URL */
+  chainApiUrls?: ChainApiUrls;
+  /** Optional mapping of chainId to API keys */
+  apiKeys?: Record<string, string>;
+  /** Optional fallback API key when chain-specific key is missing */
+  defaultApiKey?: string;
+}
+
+export class EtherscanVerifyAPIService implements WStorageService {
+  IDENTIFIER: EtherscanVerifyAPIIdentifiers;
+  private database: Database;
+
+  private readonly options: Required<EtherscanVerifyAPIServiceOptions>;
+
+  constructor(
+    identifier: EtherscanVerifyAPIIdentifiers,
+    sourcifyDatabaseService: SourcifyDatabaseService,
+    options: EtherscanVerifyAPIServiceOptions,
+  ) {
+    this.IDENTIFIER = identifier;
+    this.database = sourcifyDatabaseService.database;
+    this.options = {
+      chainApiUrls: options.chainApiUrls || {},
+      apiKeys: options.apiKeys || {},
+      defaultApiKey: options.defaultApiKey || "",
+    };
+  }
+
+  async init(): Promise<boolean> {
+    if (Object.keys(this.options.chainApiUrls).length === 0) {
+      switch (this.IDENTIFIER) {
+        case WStorageIdentifiers.EtherscanVerify:
+          this.options.chainApiUrls = await fetchEtherscanChainApiUrls();
+          break;
+        case WStorageIdentifiers.BlockscoutVerify:
+          this.options.chainApiUrls = await fetchBlockscoutChainApiUrls();
+          break;
+        case WStorageIdentifiers.RoutescanVerify:
+          this.options.chainApiUrls = await fetchRoutescanChainApiUrls();
+          break;
+        default:
+          throw new Error(`Unsupported verifier: ${this.IDENTIFIER}`);
+      }
+    }
+    return true;
+  }
+
+  async storeVerification(
+    verification: VerificationExport,
+    jobData?: {
+      verificationId: string;
+      finishTime: Date;
+    },
+  ): Promise<void> {
+    if (
+      verification.compilation.language === "Vyper" &&
+      (this.IDENTIFIER === WStorageIdentifiers.RoutescanVerify ||
+        this.IDENTIFIER === WStorageIdentifiers.BlockscoutVerify)
+    ) {
+      logger.info(
+        `${this.IDENTIFIER} does not support Etherscan API Vyper contract verification`,
+        {
+          chainId: verification.chainId,
+          address: verification.address,
+        },
+      );
+      throw new Error(
+        `${this.IDENTIFIER} does not support Etherscan API Vyper contract verification`,
+      );
+    }
+
+    const submissionContext = {
+      identifier: this.IDENTIFIER,
+      chainId: verification.chainId,
+      address: verification.address,
+    };
+
+    const apiBaseUrl = this.getApiBaseUrl(verification.chainId);
+    if (!apiBaseUrl) {
+      logger.warn(
+        "No Etherscan API endpoint configured for chain",
+        submissionContext,
+      );
```

Comment:
I think we don't need to log a warning in this case. It is totally expected that it won't run for a lot of chains. Rather change it to debug level.